Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,353 @@
+  /// * Unix:
+  ///   * `/foo/bar => /`
+  ///   * `foo/bar => nil`
+  /// * Windows:
+  ///   * `C:\foo\bar => C:\`
+  ///   * `C:foo\bar => C:`
+  ///   * `\foo\bar => \ `
+  ///   * `foo\bar => nil`
+  ///   * `\\?\UNC\server\share\file => \\?\UNC\server\share`
+  ///
+  /// TODO: Document setter. Better to document for a path what is root and what is relative.
+  ///
+  /// REVIEW NOTE: C# calls this the root. C++ implies they pull in the trailing separator (I
+  /// haven't run tests). Rust differs in that has_root is just looking for the `\` somewhere.
+  ///
+  /// REVIEW NOTE: Going with UNC or device paths not including trailing separator,
+  /// which is consistent with C# IIUC. I need to perform the experiement to verify this.
+  ///
+  /// REVIEW NOTE: I'm going with component as return type, which implies the more common
+  /// use would be to read or inspect the root, rather than forming new paths from the same root.
+  /// For that use case, you can do `FilePath(otherPath.root)`, or use
+  /// the `relativePath`'s setter, or use the `/` operator (if we go that route).
+  public var root: FilePath.Component? {
+    get {
+      let rootEnd = _parseRoot().0
+      guard rootEnd != storage.startIndex else { return nil }
+      return Component(self, ..<rootEnd)
+    }
+    set {
+      fatalError()
+    }
+  }
+
+  /// REVIEW NOTE: I'm dropping rootName and rootDirectory. I think
+  /// C++ was a little misguided here after looking more at C#. I think
+  /// a better alternative would be to actually identify and work with
+  /// the root when on Windows, more akin to Rust's `Prefix` type.
+
+  /// Gets or sets the relative portion of the path (everything after root)
+  ///
+  /// Examples:
+  /// * Unix:
+  ///   * `/foo/bar => foo/bar`
+  ///   * `foo/bar => foo/bar`
+  ///   * `/ => ""`
+  /// * Windows:
+  ///   * `C:\foo\bar => foo\bar`
+  ///   * `foo\bar => foo\bar`
+  ///   * `\\?\UNC\server\share\file => file`
+  ///   * `\ => ""`
+  ///
+  /// REVIEW NOTE: Not optional, can return empty paths. Empty paths
+  /// are relative (but are not roots), hence non-optional.
+  public var relativePath: FilePath {
+    get {
+      let (_, relIdx) = _parseRoot()
+      return FilePath(slice: storage[relIdx...])
+    }
+    set { fatalError() }
+  }
+
+  /// Returns the final component of the path. If the path is a normal file, this
+  /// will be the file name. If the path is a directory, this is the directory name.
+  /// Returns `nil` if the path is empty or only contains a root
+  ///
+  /// REVIEW NOTE: We follow  `basename` semantics, unlike `file_name` in
+  /// C++ which treats trailing `/` specially, or in Rust which returns nil for `..`
+  ///
+  /// REVIEW NOTE: shell and C basename will return `/` if the path is `/`. I don't
+  /// think we want to, it seems clearer that this only is in the relative portion of the path,
+  /// and doing so would make the setter weird.
+  public var basename: FilePath.Component? {
```

Comment:
I understand the reasoning for making this optional, but I think it makes the API fairly unergonomic for clients.  Pretty much every client that wasn't to know the last path component would need to check for it and throw an error.  Maybe that's the correct behavior, but it seems a bit tedious.

Reply:
It's always a debate whether presence or non-presence of a result should be elevated to the type system rather than a convention on a particular return value. If it's salient we like to promote it to the type level. Paths can be empty, but components generally cannot (at least if they were extracted from a path). The empty path has no components, rather than it having a single empty component. Inserting an empty component is a weird thing to do and indicative of a potential bug, so we try to avoid empty components. We may even want to forbid empty components.

`relativePath` up above and `dirname` below uses empty paths because every path has a `dirname` and `relativePath`, which could be empty. `root` is optional not only because the presence or non-presence is highly meaningful, but also because it is a component. These component APIs also have `setters`, where it is useful to set a component to `nil`, which I feel is clearer algebraically than `""`.